lib/services/cloud-service.ts

```diff
@@ -1,89 +1,108 @@
 import * as path from "path";
 import { EventEmitter } from "events";
-import promiseRetry = require("promise-retry");
-
-export abstract class CloudService extends EventEmitter {
-	private static OPERATION_STATUS_CHECK_INTERVAL = 1500;
-	private static OPERATION_STATUS_CHECK_RETRY_COUNT = 8;
-	private static OPERATION_COMPLETE_STATUS = "Success";
-	private static OPERATION_FAILED_STATUS = "Failed";
-	// TODO: Remove after 10.05.2018
-	private static OPERATION_IN_PROGRESS_STATUS_OLD = "Building";
-	private static OPERATION_IN_PROGRESS_STATUS = "InProgress";
-
-	protected outputCursorPosition: number;
+import { v4 } from "uuid";
+
+import { CloudOperationMessageTypes, CloudCommunicationEvents } from "../constants";
+
+export abstract class CloudService extends EventEmitter implements ICloudOperationService {
+	private static readonly CLOUD_OPERATION_VERSION_1: string = "v1";
 	protected abstract failedToStartError: string;
 	protected abstract failedError: string;
-	protected abstract getServerResults(serverResult: IBuildServerResult): IServerItem[];
-	protected abstract getServerLogs(logsUrl: string, buildId: string): Promise<void>;
 
-	public abstract getServerOperationOutputDirectory(options: IOutputDirectoryOptions): string;
+	private cloudOperations: IDictionary<ICloudOperation>;
 
-	constructor(protected $fs: IFileSystem,
+	constructor(protected $errors: IErrors,
+		protected $fs: IFileSystem,
 		protected $httpClient: Server.IHttpClient,
-		protected $logger: ILogger) {
+		protected $logger: ILogger,
+		private $nsCloudOperationFactory: ICloudOperationFactory,
+		private $nsCloudS3Service: IS3Service,
+		private $nsCloudOutputFilter: ICloudOutputFilter,
+		private $processService: IProcessService) {
 		super();
+		this.cloudOperations = Object.create(null);
+		this.$processService.attachToProcessExitSignals(this, this.cleanup.bind(this));
 	}
 
-	protected async getObjectFromS3File<T>(pathToFile: string): Promise<T> {
-		return JSON.parse(await this.getContentOfS3File(pathToFile));
+	public abstract getServerOperationOutputDirectory(options: IOutputDirectoryOptions): string;
```

Comment:
Can we make getServerOperationOutputDirectory and getServerResults virtual - returning "" and [] respectively. If I am correct, these are not required by all services

Reply:
Done.